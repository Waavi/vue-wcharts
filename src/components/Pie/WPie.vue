<template>
    <g
        v-if="visible"
        :style="stylesCmp"
    >
        <foreignObject :style="contentStyles">
            <slot :values="values" />
        </foreignObject>

        <path
            v-for="(path, i) in paths"
            :id="i"
            :key="i"
            :d="path.d"
            :fill="path.fill"
            :stroke="path.stroke"
            :style="{
                ...pathStylesCmp,
                opacity: activePath === null ? 1 : activePath === i ? 1 : opacityDisabled,
            }"
            v-on="pieListeners"
        />
    </g>
</template>

<script>
import VueTypes from 'vue-types'
import omit from 'lodash.omit'
import noop from 'lodash.noop'
import merge from 'lodash.merge'
import pie from 'd3-shape/src/pie'
import arc from 'd3-shape/src/arc'
import themeMixin from '../../mixins/theme'

export default {
    name: 'WPie',
    type: 'pie',
    inject: ['Chart'],
    mixins: [themeMixin],
    props: {
        // internal props set by the parent (WPieChart)
        index: VueTypes.number,
        datakey: VueTypes.string.isRequired,
        trigger: VueTypes.oneOf(['hover', 'click', 'manual']).def('hover'),
        angles: VueTypes.oneOfType([
            VueTypes.number,
            VueTypes.arrayOf(VueTypes.number).def([0, Math.PI * 2]),
        ]).def([0, Math.PI * 2]),
        radius: VueTypes.oneOfType([
            VueTypes.number,
            VueTypes.arrayOf(VueTypes.number).def([0, 100]),
        ]).def([0, 100]),
        styles: VueTypes.object,
        pathStyles: VueTypes.shape({
            stroke: VueTypes.string,
        }).loose,
        opacityDisabled: VueTypes.number.def(0.5),
        active: VueTypes.oneOfType([Number, null]),
    },
    data () {
        return {
            activePath: null,
        }
    },
    computed: {
        stylesCmp () {
            return {
                ...this.themeStyles.styles,
                ...this.styles,
            }
        },
        pathStylesCmp () {
            return {
                ...omit(this.themeStyles.path, ['stroke']),
                ...omit(this.pathStyles, ['stroke']),
                ...(this.trigger === 'click' ? { cursor: 'pointer' } : {}),
            }
        },
        // Active elem
        visible () {
            return this.Chart.activeElements.includes(this.index)
        },
        curRadius () {
            const innerRadius = Array.isArray(this.radius) ? this.radius[0] : 0
            const outerRadius = Array.isArray(this.radius) ? this.radius[1] : 100

            return [innerRadius, outerRadius]
        },
        // Generate angles by number or array
        curAngles () {
            const { angles } = this
            return Array.isArray(angles) ? angles : [0, angles]
        },
        // Values
        curValues () {
            return this.Chart.data.map(item => item[this.datakey])
        },
        // Values
        values () {
            const total = this.curValues.reduce((acc, a) => acc + a, 0)
            return this.Chart.data.map((item, index) => ({
                ...item,
                percentage: item[this.datakey] * 100 / total,
                color: this.Chart.colors[index],
            }))
        },
        // Generate arrry of arcs
        arcs () {
            return pie()
                .startAngle(this.curAngles[0])
                .endAngle(this.curAngles[1])
                .sortValues(noop)(this.curValues)
        },
        // Return the path of angles
        draw () {
            return arc()
                .innerRadius(this.curRadius[0])
                .outerRadius(this.curRadius[1])
        },
        // Generate paths array by arcs
        paths () {
            return this.arcs.map(this.draw).map((d, index) => ({
                d,
                fill: this.Chart.colors[index],
                stroke: this.pathStylesCmp.stroke,
            }))
        },
        // Event Listeners
        pieListeners () {
            return merge({}, this.$listeners, {
                click: (event) => {
                    if (this.trigger === 'click') {
                        this.setActivePath(parseInt(event.target.id, 0))
                        this.handleActive(event)
                    }
                    this.$emit('onClick')
                },
                mouseenter: (event) => {
                    if (this.trigger === 'hover') {
                        this.setActivePath(parseInt(event.target.id, 0))
                        this.handleActive(event)
                    }
                    this.$emit('onMouseenter')
                },
                mouseleave: () => {
                    if (['hover', 'click'].includes(this.trigger)) {
                        this.setActivePath(null)
                        this.Chart.cleanActive()
                    }
                    this.$emit('onMouseleave')
                },
            })
        },
        // Slot styles
        contentStyles () {
            const [, width] = this.curRadius
            const size = `${width * 2}px`
            return {
                width: size,
                height: size,
                transform: `translate(-${width}px, -${width}px)`,
            }
        },
    },
    watch: {
        active (newValue) {
            this.activePath = newValue
        },
    },
    methods: {
        setActivePath (id) {
            if ([null, undefined].includes(this.active)) this.activePath = id
        },
        handleActive (event) {
            const { id } = event.target
            const { setActive, colors } = this.Chart
            const value = this.curValues[id]
            const color = this.fill || colors[id]
            const el = {
                id,
                value: [{
                    value,
                    color,
                }],
            }
            // Set active bar to show tooltip
            setActive(el, event)
        },
    },
}
</script>
