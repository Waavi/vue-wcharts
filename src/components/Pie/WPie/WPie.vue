<template>
    <g
        v-if="active"
        class="pie-sectors"
    >
        <foreignObject
            v-if="slots"
            :style="contentStyles"
        >
            <slot />
        </foreignObject>

        <g
            v-for="(path, i) in paths"
            :key="i"
            class="pie-sector"
        >
            <path
                :id="i"
                :d="path.d"
                :fill="path.fill"
                :stroke="path.stroke"
                :style="{
                    ...pathStylesCmp,
                    opacity: activePath === null ? 1 : activePath === i ? 1 : opacityDisabled,
                }"
                v-on="pieListeners"
            />
        </g>
    </g>
</template>

<script>
import VueTypes from 'vue-types'
import omit from 'lodash.omit'
import TweenLite from 'gsap/TweenLite'
import merge from '../../../utils/merge'
import themeMixin from '../../../mixins/theme'
import visibleMixin from '../../../mixins/visible'
import { isNumber } from '../../../utils/checks'
import {
    getSectorPath, mathSign, parseDeltaAngle,
} from '../../../utils/mathsPie'

export default {
    name: 'WPie',
    type: 'pie',
    inject: ['Chart'],
    mixins: [themeMixin, visibleMixin],
    props: {
        // Internal props set by the parent (WPieChart)
        index: VueTypes.number,
        datakey: VueTypes.string,
        trigger: VueTypes.oneOf(['hover', 'click', 'manual']).def('hover'),
        startAngle: VueTypes.number.def(-1),
        endAngle: VueTypes.number.def(-1),
        radius: VueTypes.oneOfType([
            VueTypes.number,
            VueTypes.arrayOf(VueTypes.number).def([0, 100]),
        ]).def([0, 100]),
        maxValue: VueTypes.number.def(-1), // Max value to limit curValues
        // Styles
        styles: VueTypes.object,
        pathStyles: VueTypes.shape({
            stroke: VueTypes.string,
        }).loose,
        color: VueTypes.string,
        borderRadius: VueTypes.oneOfType([VueTypes.number, VueTypes.string]).def(0),
        opacityDisabled: VueTypes.number.def(0.5),
        // Animation
        animation: VueTypes.bool.def(true),
        animationDuration: VueTypes.number.def(2.5),
        // index of item active
        itemActive: VueTypes.oneOfType([Number, null]),
    },
    data () {
        return {
            activePath: null,
            animatedSectors: [],
        }
    },
    computed: {
        // Active elem
        active () {
            return this.Chart.activeElements.includes(this.index)
        },
        // Radius
        curRadius () {
            const innerRadius = Array.isArray(this.radius) ? this.radius[0] : 0
            const outerRadius = Array.isArray(this.radius) ? this.radius[1] : 100

            return { innerRadius, outerRadius }
        },
        // Values
        curValues () {
            return this.Chart.data.map(item => (isNumber(item[this.datakey]) ? item[this.datakey] : this.maxValue))
        },
        // Sectors
        sectors () {
            const { paddingAngle, curCx, curCy } = this.Chart
            const {
                curValues, curRadius, maxValue, borderRadius,
            } = this
            const startAngle = this.startAngle >= 0 ? this.startAngle : this.Chart.startAngle
            const endAngle = this.endAngle >= 0 ? this.endAngle : this.Chart.endAngle
            const sum = maxValue >= 0 ? maxValue : this.curValues.reduce((acc, a) => acc + a, 0)
            let prev

            const { length } = curValues
            const deltaAngle = parseDeltaAngle({ startAngle, endAngle })
            const absDeltaAngle = Math.abs(deltaAngle)
            const totalPadingAngle = (absDeltaAngle >= 360 ? length : (length - 1)) * paddingAngle
            const realTotalAngle = absDeltaAngle - totalPadingAngle

            return curValues.map((value, index) => {
                let tempStartAngle
                const percentage = (isNumber(value) ? value : 0) / sum

                if (index) {
                    tempStartAngle = prev.endAngle + mathSign(deltaAngle) * paddingAngle
                } else {
                    tempStartAngle = startAngle
                }
                const tempEndAngle = tempStartAngle + mathSign(deltaAngle) * (percentage * realTotalAngle)

                prev = {
                    cx: curCx,
                    cy: curCy,
                    value,
                    percentage,
                    startAngle: tempStartAngle,
                    endAngle: tempEndAngle,
                    borderRadius,
                    ...curRadius,
                }

                return prev
            })
        },
        // Generate paths array by arcs
        paths () {
            const sectors = (this.animation ? this.animatedSectors : this.sectors)
            return sectors.map((sector, index) => ({
                d: getSectorPath(sector),
                fill: this.color || this.Chart.colors[index],
                stroke: this.pathStylesCmp.stroke,
            }))
        },
        // Event Listeners
        pieListeners () {
            return merge({}, this.$listeners, {
                click: (event) => {
                    if (this.trigger === 'click') {
                        this.setActivePath(parseInt(event.target.id, 0))
                        this.handleActive(event)
                    }
                    this.$emit('onClick')
                },
                mouseenter: (event) => {
                    if (this.trigger === 'hover') {
                        this.setActivePath(parseInt(event.target.id, 0))
                        this.handleActive(event)
                    }
                    this.$emit('onMouseenter')
                },
                mouseleave: () => {
                    if (['hover', 'click'].includes(this.trigger)) {
                        this.setActivePath(null)
                        this.Chart.cleanActive()
                    }
                    this.$emit('onMouseleave')
                },
            })
        },
        // Styles
        contentStyles () {
            const { width, height } = this.Chart
            const { outerRadius } = this.curRadius
            const size = `${outerRadius * 2}px`
            return {
                width: size,
                height: size,
                transform: `translate(${width / 2 - outerRadius}px, ${height / 2 - outerRadius}px)`,
                overflow: 'initial',
            }
        },
        stylesCmp () {
            return {
                ...this.themeStyles.styles,
                ...this.styles,
            }
        },
        pathStylesCmp () {
            return {
                ...omit(this.themeStyles.path, ['stroke']),
                ...omit(this.pathStyles, ['stroke']),
                ...(this.trigger === 'click' ? { cursor: 'pointer' } : {}),
            }
        },
        // Slots
        slots () {
            return Object.keys(this.$slots).length
        },
    },
    watch: {
        sectors: {
            handler (sectors) {
                if (this.animation) {
                    // Initialize sectors
                    const { startAngle } = this
                    this.animatedSectors = sectors.map(s => ({ ...s, startAngle, endAngle: startAngle }))
                    // Interpolate angles
                    sectors.forEach((s, index) => {
                        TweenLite.to(this.animatedSectors[index], this.animationDuration, {
                            startAngle: sectors[index].startAngle,
                            endAngle: sectors[index].endAngle,
                        })
                    })
                }
            },
            immediate: true,
        },
        itemActive: {
            handler (id) {
                if (id) this.activePath = id
            },
            immediate: true,
        },
    },
    methods: {
        setActivePath (id) {
            if ([null, undefined].includes(this.itemActive)) this.activePath = id
        },
        handleActive (event) {
            const { id } = event.target
            const { setActive, colors } = this.Chart
            const datum = this.Chart.data[id]
            const color = this.fill || colors[id]
            const value = [{ value: this.curValues[id], color }]
            // Set active bar to show tooltip
            setActive({
                id, value, datum,
            }, event)
        },
    },
}
</script>
