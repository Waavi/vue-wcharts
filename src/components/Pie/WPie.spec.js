import { shallowMount, mount } from '@vue/test-utils'
import WPie from './WPie.vue'

describe('Components/WPie', () => {
    const propsData = {
        datakey: 'one',
        index: 0,
    }
    const dataset = [
        {
            name: 'Page A', one: 1000, two: 2400, three: 2400,
        },
        {
            name: 'Page B', one: 3000, two: 1398, three: 2210,
        },
        {
            name: 'Page C', one: 2000, two: 9800, three: 2290,
        },
        {
            name: 'Page D', one: 2780, two: 3908, three: 2000,
        },
        {
            name: 'Page E', one: 1890, two: 0, three: 1700,
        },
        {
            name: 'Page F', one: 2390, two: 3800, three: 2500,
        },
        {
            name: 'Page G', one: 3490, two: 4300, three: -1000,
        },
    ]

    const provide = {
        Chart: {
            dataset,
            data: dataset,
            width: 740,
            height: 400,
            cx: '50%',
            cy: '50%',
            startAngle: 0,
            endAngle: 360,
            paddingAngle: 0,
            activeElements: [0],
            colors: ['red', 'red', 'red', 'red', 'red', 'red', 'red'],
            curCx: 300,
            curCy: 200,
            setActive: () => undefined,
            cleanActive: () => undefined,
        },
    }

    const defaultConfig = {
        propsData,
        provide,
    }

    it(`Should be render correctly`, () => {
        const wrapper = shallowMount(WPie, defaultConfig)
        expect(wrapper.html()).toMatchSnapshot()
    })

    it('It emits the handleClick event', () => {
        const wrapper = mount(WPie, defaultConfig)
        wrapper.findAll('path').at(0).trigger('click')
        expect(wrapper.emitted('onClick')).toHaveLength(1)
    })

    it('It emits the handleMouseEnter  event', () => {
        const wrapper = mount(WPie, defaultConfig)
        wrapper.findAll('path').at(0).trigger('mouseenter')
        expect(wrapper.emitted('onMouseenter')).toHaveLength(1)
    })

    it('Should be active 1', () => {
        const wrapper = mount(WPie, defaultConfig)
        wrapper.setProps({ itemActive: 1 })
        expect(wrapper.props('itemActive')).toBe(1)
        expect(wrapper.vm.activePath).toBe(1)
    })
})
