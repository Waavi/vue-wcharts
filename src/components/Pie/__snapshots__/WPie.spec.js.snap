// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/WPie Should be render correctly 1`] = `
<g class="pie-sectors">
  <foreignObject style="width: 200px; height: 200px; transform: translate(100px, 100px);"></foreignObject>
  <g class="pie-sector">
    <path id="0" d="M 400,200
    A 100,100,0,
    0,0,
    392.8794897808724,162.94058313943916
L 300,200 Z" fill="red" style="opacity: 1;"></path>
  </g>
  <g class="pie-sector">
    <path id="1" d="M 392.8794897808724,162.94058313943916
    A 100,100,0,
    0,0,
    305.2177985437235,100.13621988750317
L 300,200 Z" fill="red" style="opacity: 1;"></path>
  </g>
  <g class="pie-sector">
    <path id="2" d="M 305.2177985437235,100.13621988750317
    A 100,100,0,
    0,0,
    235.03715405133886,123.97481571051262
L 300,200 Z" fill="red" style="opacity: 1;"></path>
  </g>
  <g class="pie-sector">
    <path id="3" d="M 235.03715405133886,123.97481571051262
    A 100,100,0,
    0,0,
    201.83225448497927,219.05501877454145
L 300,200 Z" fill="red" style="opacity: 1;"></path>
  </g>
  <g class="pie-sector">
    <path id="4" d="M 201.83225448497927,219.05501877454145
    A 100,100,0,
    0,0,
    238.56690325751902,278.90484538118665
L 300,200 Z" fill="red" style="opacity: 1;"></path>
  </g>
  <g class="pie-sector">
    <path id="5" d="M 238.56690325751902,278.90484538118665
    A 100,100,0,
    0,0,
    324.33541871316555,296.99374926279995
L 300,200 Z" fill="red" style="opacity: 1;"></path>
  </g>
  <g class="pie-sector">
    <path id="6" d="M 324.33541871316555,296.99374926279995
    A 100,100,0,
    0,0,
    400,200.00000000000003
L 300,200 Z" fill="red" style="opacity: 1;"></path>
  </g>
</g>
`;

exports[`Components/WPie Should be render correctly with custom radius 1`] = `
<g class="pie-sectors">
  <foreignObject style="width: 360px; height: 360px; transform: translate(180px, 20px);"></foreignObject>
  <g class="pie-sector">
    <path id="0" d="M 480,200
    A 180,180,0,
    0,0,
    467.1830816055703,133.29304965099047
L 439.31923467130855,144.41087470915872
            A 150,150,0,
            0,1,
            450,200 Z" fill="red" style="opacity: 1;"></path>
  </g>
  <g class="pie-sector">
    <path id="1" d="M 467.1830816055703,133.29304965099047
    A 180,180,0,
    0,0,
    309.39203737870235,20.245195797505716
L 307.8266978155853,50.20432983125477
            A 150,150,0,
            0,1,
            439.31923467130855,144.41087470915872 Z" fill="red" style="opacity: 1;"></path>
  </g>
  <g class="pie-sector">
    <path id="2" d="M 309.39203737870235,20.245195797505716
    A 180,180,0,
    0,0,
    183.0668772924099,63.15466827892271
L 202.55573107700826,85.96222356576894
            A 150,150,0,
            0,1,
            307.8266978155853,50.20432983125477 Z" fill="red" style="opacity: 1;"></path>
  </g>
  <g class="pie-sector">
    <path id="3" d="M 183.0668772924099,63.15466827892271
    A 180,180,0,
    0,0,
    123.29805807296268,234.29903379417465
L 152.7483817274689,228.5825281618122
            A 150,150,0,
            0,1,
            202.55573107700826,85.96222356576894 Z" fill="red" style="opacity: 1;"></path>
  </g>
  <g class="pie-sector">
    <path id="4" d="M 123.29805807296268,234.29903379417465
    A 180,180,0,
    0,0,
    189.42042586353426,342.028721686136
L 207.85035488627852,318.35726807178
            A 150,150,0,
            0,1,
            152.7483817274689,228.5825281618122 Z" fill="red" style="opacity: 1;"></path>
  </g>
  <g class="pie-sector">
    <path id="5" d="M 189.42042586353426,342.028721686136
    A 180,180,0,
    0,0,
    343.803753683698,374.58874867303996
L 336.5031280697483,345.4906238942
            A 150,150,0,
            0,1,
            207.85035488627852,318.35726807178 Z" fill="red" style="opacity: 1;"></path>
  </g>
  <g class="pie-sector">
    <path id="6" d="M 343.803753683698,374.58874867303996
    A 180,180,0,
    0,0,
    480,200.00000000000006
L 450,200.00000000000003
            A 150,150,0,
            0,1,
            336.5031280697483,345.4906238942 Z" fill="red" style="opacity: 1;"></path>
  </g>
</g>
`;
